keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+object contains the template for a Pod that encodes a semantic readiness check for the components of an Application. 
+The requirements of implementing a ApplicationHealthCheck Pod are as follows.
+
+1. A health check Pod must implement a readiness check.
+1. The readiness check of the Pod may only return successfully when the conditions of readiness are met for the entire 
+application. 
+1. An ApplicationHealthCheck Pod should not supply other application telemetry. When its readiness check fails that 
+telemetry would be unavailable (at precisely the time it is most necessary).
+
+### Application Controller
+The Application Controller will be responsible for configuring ownership to enable GC (Garbage Collection) for an 
+Application's components and for updating the `Status` of the Application with respect to installation 
+status of the components and dependencies of the Application and to the readiness of the Application's health check.
+
+#### Garbage Collection
+In order to facilitate GC for Application components, an  `OwnerRefernce` from the API object must be 
+added to all of its indicated component objects. As an Application can be composed of arbitrary Kinds, like the 
+Garbage Collector, the Application Controller will have to use the discovery API to determine when new Kinds are 
+registered with the API server. It will have to watch all Kinds that may be included in an Application's `Components`.
+
+When the Application Controller detects a newly created or updated object it shall do the following.
+1. [Retrieve the Application for the object](#retrieving-applications).
+1. If the Application's `Selector` matches the object, and if the object does not contain an `OwnerReference` to the 
+Application, the Application Controller shall add an  `OwnerReference` from the Application to the object.
+
+As there is no guarantee that an Application's components will be created prior to the Application, the Application 
+Controller will do the following upon creation of an Application.
+
+1. [Select the Application's components](#selecting-components).
+1. For all selected objects, the Application Controller will add an `OwnerReference` from the Application to the object.
+
+When an Application object is deleted, the Garbage Collector will detect that the owner of the Application's components 
+no longer exists at the API Server. This will trigger deletion of the children objects. Note that the Application 
+Controller does not establish an ownership relationship between an Application and its dependencies. This is by design.
+Also note that, if the Application is composed of Kubernetes primitives that, in turn, aggregate other objects, it is 
+sufficient to only add an OwnerReference to the top level object. For instance, deleting a Application containing a 
+Deployment will delete the Deployment's ReplicaSets and Pods.
```

Comment:
I'm not totally sure that this is true with CronJob and the Job objects that it creates... but I think that might be an issue with CronJobs, and not something we'd have to cover here.

Reply:
I believe deletion of CronJobs does cascade delete Jobs (which cascade deletes Pods).